pilot/pkg/networking/core/v1alpha3/cluster.go

```diff
@@ -799,18 +799,20 @@ func applyOutlierDetection(cluster *apiv2.Cluster, outlier *networking.OutlierDe
 }
 
 func applyLoadBalancer(cluster *apiv2.Cluster, lb *networking.LoadBalancerSettings, port *model.Port, proxy *model.Proxy, meshConfig *meshconfig.MeshConfig) {
+	lbSetting := loadbalancer.GetLocalityLbSetting(meshConfig.GetLocalityLbSetting(), lb.GetLocalityLbSetting())
 	if cluster.OutlierDetection != nil {
 		if cluster.CommonLbConfig == nil {
 			cluster.CommonLbConfig = &apiv2.Cluster_CommonLbConfig{}
 		}
-		// Locality weighted load balancing
-		cluster.CommonLbConfig.LocalityConfigSpecifier = &apiv2.Cluster_CommonLbConfig_LocalityWeightedLbConfig_{
```

Comment:
Yes, no side effect. this is why i always set it here before

Reply:
But, I think there is no harm in adding this check also right? I prefer add it just for safety.

Reply:
yep

Reply:
I think it matters here 
```
load_balancing_weight
(UInt32Value) Optional: Per priority/region/zone/sub_zone weight; at least 1. The load balancing weight for a locality is divided by the sum of the weights of all localities at the same priority level to produce the effective percentage of traffic for the locality.

Locality weights are only considered when locality weighted load balancing is configured. These weights are ignored otherwise. If no weights are specified when locality weighted load balancing is enabled, the locality is assigned no load.
```

Reply:
Without this, does each group of endpoints within one locality has the same weight as the other locality?

Reply:
But without locality enabled, we do not even calculate weights for EDS endpoints https://github.com/istio/istio/blob/master/pilot/pkg/proxy/envoy/v2/eds.go#L410

Reply:
and same for clusters - if it is not enabled `localityLB` is nil - so we do not calculate weights. 
https://github.com/istio/istio/blob/master/pilot/pkg/networking/core/v1alpha3/cluster.go#L872

Reply:
Have a look at `buildLocalityLbEndpointsFromShards`, 
```
	for _, locLbEps := range localityEpMap {
		var weight uint32
		for _, ep := range locLbEps.LbEndpoints {
			weight += ep.LoadBalancingWeight.GetValue()
		}
		locLbEps.LoadBalancingWeight = &wrappers.UInt32Value{
			Value: weight,
		}
		locEps = append(locEps, locLbEps)
	}
```

Reply:
Ok. That makes sense. But I did not understand why we have to calculate weights when locality is not enabled.

Also "**Locality weights are only considered when locality weighted load balancing is configured.These weights are ignored otherwise.**" - If we do not enable cluster for locality lb,  the weights calculated above is ignored and traffic is routed to all endpoints based on the LB set - Is n't that what we want if locality is disabled? Why should we calculate some weights and enable cluster for locality lb ?

Reply:
The main target is to make traffic evenly distributed to each endpoint no matter where it resides.

I am not sure when both the weight and the locality config unset, does envoy treat each endpoint across different localities fairly? Or treat each locality group fairly.  So in order to simplify, always set both of them